helix-core/src/main/java/org/apache/helix/controller/stages/CustomizedStateOutput.java

```diff
@@ -0,0 +1,135 @@
+    _customizedStateMetaMap = new HashMap<>();
+  }
+
+  public void setCustomizedState(String stateType, String resourceName, Partition partition,
+      String instanceName, String state) {
+    if (!_customizedStateMap.containsKey(stateType)) {
+      _customizedStateMap.put(stateType,
+          new HashMap<String, Map<Partition, Map<String, String>>>());
+    }
+    if (!_customizedStateMap.get(stateType).containsKey(resourceName)) {
+      _customizedStateMap.get(stateType).put(resourceName,
+          new HashMap<Partition, Map<String, String>>());
+    }
+    if (!_customizedStateMap.get(stateType).get(resourceName).containsKey(partition)) {
+      _customizedStateMap.get(stateType).get(resourceName).put(partition,
+          new HashMap<String, String>());
+    }
+    _customizedStateMap.get(stateType).get(resourceName).get(partition).put(instanceName, state);
+  }
+
+  /**
+   * given (stateType, resource, partition, instance), returns customized state
+   * @param stateType
+   * @param resourceName
+   * @param partition
+   * @param instanceName
+   * @return
+   */
+  public String getCustomizedState(String stateType, String resourceName, Partition partition,
+      String instanceName) {
+    Map<String, Map<Partition, Map<String, String>>> map = _customizedStateMap.get(stateType);
+    if (map != null) {
+      Map<Partition, Map<String, String>> resourceMap = map.get(resourceName);
+      if (resourceMap != null) {
+        Map<String, String> instanceStateMap = resourceMap.get(partition);
+        if (instanceStateMap != null) {
+          return instanceStateMap.get(instanceName);
+        }
+      }
+    }
+    return null;
+  }
+
+  /**
+   * Given stateType, returns resource customized state map (resource -> parition -> instance ->
+   * customizedState)
+   * @param stateType
+   * @return
+   */
+  public Map<String, Map<Partition, Map<String, String>>> getResourceCustomizedStateMap(
+      String stateType) {
+    if (_customizedStateMap.containsKey(stateType)) {
+      return _customizedStateMap.get(stateType);
+    }
+    return Collections.emptyMap();
+  }
+
+  /**
+   * given (stateType, resource), returns (partition -> instance-> customizedState) map
+   * @param stateType
+   * @param resourceName
+   * @return
+   */
+  public Map<Partition, Map<String, String>> getCustomizedStateMap(String stateType,
+      String resourceName) {
+    if (_customizedStateMap.containsKey(stateType)) {
+      Map<String, Map<Partition, Map<String, String>>> map = _customizedStateMap.get(stateType);
+      if (map.containsKey(resourceName)) {
+        return map.get(resourceName);
+      }
+    }
+    return Collections.emptyMap();
+  }
+
+  /**
+   * given (stateType, resource, partition), returns (instance-> customizedState) map
+   * @param stateType
+   * @param resourceName
+   * @param partition
+   * @return
+   */
+  public Map<String, String> getCustomizedStateMap(String stateType, String resourceName,
```

Comment:
This method should call the other getCustomizedStateMap() to avoid duplicate code.
BTW, can we have a better name to differentiate?